otri/validation/__init__.py

```diff
@@ -0,0 +1,572 @@
+    if key not in data.keys():
+        data[key] = list()
+    data[key].append(repr(label))
+
+
+class ValidatorFilter(Filter):
+
+    '''
+    This Filter is used to apply a check to a list of atoms.
+    This is an abstract class and should be further extended implementing at least, the `_check`,
+    `_on_ok` and `_on_error` methods.
+
+    The `_check` method should return some subclass of `AtomException` in case of some kind of
+    problem with the atom's data. If the method does not return any value, the atom is assumed
+    to be ok, and passed on. Any value can be returned.
+    '''
+
+    def _on_data(self, data: Mapping, index: int):
+        '''
+        Called when input data is found.
+
+        Parameters:
+            data : Mapping
+                Popped data from an input.
+
+            index : int
+                The index of the input the data has been popped from.
+        '''
+        result = self._check(data)
+        if result is None:
+            self._on_ok(data, index)
+        else:
+            log.v(msg="{}. Data: {}\nAnalysis: {}.".format(self, data, result))
+            self._on_result(data, result, index)
+
+    def _check(self, data: Mapping) -> Union[Any, None]:
+        '''
+        Check a single atom. Must be implemented. Should return None if the atom is ok or some value
+        indicating the problem. See `AtomException` for examples.
+
+        Parameters:
+            data : Mapping
+                The data to check.
+        '''
+        raise NotImplementedError("ValidatorFilter is an abstract class, please extend it.")
+
+    def _on_ok(self, data: Mapping, index: int):
+        '''
+        Called if data resulted ok.
+
+        Parameters:
+            data : Mapping
+                The checked data.\n
+            index : int
+                The index of the input the data has been popped from.
+        '''
+        raise NotImplementedError("ValidatorFilter is an abstract class, please extend it.")
+
+    def _on_result(self, data: Mapping, result: Any, index: int):
+        '''
+        Called if something was returned by `_check`.
+
+        Parameters:
+            data : Mapping
+                The checked data.
+
+            result : Any
+                The result of the analysis.
+
+            index : int
+                The index of the input the data has been popped from.
+        '''
+        raise NotImplementedError("ValidatorFilter is an abstract class, please extend it.")
+
+    # This separation of the method from the class allows to override it globally, on class level or
+    # on object level, as needed.
+
+    @classmethod
+    def _add_label(cls, data: Mapping, label: Any):
+        '''
+        How the `ValidatorFilter` handles appending a label to an atom. This uses directly the
+        `append_label` method of this module.
+        '''
+        append_label(data, label)
+
+
+class LinearValidator(ValidatorFilter):
```

Comment:
Codacy found an issue: [Method '_check' is abstract in class 'ValidatorFilter' but is not overridden](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6277039)